src/CakeSlayer/Has.hs

```diff
@@ -11,15 +8,31 @@ module CakeSlayer.Has
 
 {- | General type class representing which @field@ is in @env@.
 
-Instead of plain usage like this:
+Assuming that you have the following environment data type:
+
+@
+data Env = Env
+    { envJwtSecret :: !JwtSecret
+    , ...
+    }
+@
+
+Then you should first write instances for each field in the following manner:
+
+@
+instance Has JwtSecret Env where
+    obtain = envJwtSecret
```

Comment:
Should we mention `{-# INLINE #-}` pragmas here?

Also, I've discovered recently, that GHC automatically generates instances of the following `HasField` typeclass.

* https://hackage.haskell.org/package/base-4.12.0.0/docs/GHC-Records.html#t:HasField

I feel like we can improve our `Has` pattern by using this approach so users need to write less code. Would you be interested in hearing my ideas? :slightly_smiling_face:

Reply:
Yes, sure, go ahead!

Reply:
So, if you have a data type like this:

```haskell
data Env = Env
    { envInt :: !Int
    , envString :: !String
    }
```

GHC will generate the following instances automatically:

```haskell
instance HasField "envInt" Env Int where
  getField :: Env -> Int
  getField = envInt

instance HasField "envString" Env String where
  getField :: Env -> String
  getField = envString
```

You can have now use `getField @"envInt"` instead of `envInt`. Well, the fact that you can doesn't mean that you should in real code. But since GHC already generates overloaded getters, I believe it's possible to improve our interface and remove some boilerplate. So instead of writing the following code:

```haskell
data Env = Env
    { envInt :: !Int
    , envString :: !String
    }

instance Has Int Env where
    obtain = envInt
    {-# INLINE obtain #-}

instance Has String Env where
    obtain = envString
    {-# INLINE obtain #-}
```

It should be possible to write the following:

```haskell
data Env = Env
    { envInt :: !Int
    , envString :: !String
    } deriving Has Int Env via Field "envInt" Int Env
      deriving Has String Env via Field "envString" String Env
```

This is just a sketch of an idea. I'm not sure that it's possible but It seems possible though I haven't tried to do this. If you think that this is a good idea and if you want to play with DerivingVia even more, you can try to implement  it :slightly_smiling_face:  But if we can implement this, it will mean that our `Has` approach is the coolest `Has` approach, it's unique, smart, fancy and convenient.

Reply:
That's incredible :fire:  
I've created #14 to track this, because I guess we need a bit of time to check and play with it :slightly_smiling_face: